### PR TITLE
[Draft] Add support for P2WPKh (Segwit) and P2TR key-path (Taproot)

### PR DESCRIPTION
**Pull request by lamafab**
Related to https://github.com/trustwallet/wallet-core/issues/3178

Although testing and debugging for BRC20 are still in progress, we plan to add support for P2WPKH and P2TR key-path earlier. Personally, I'm not a fan of merging this prematurely. However, here's the current checklist:

- [x] Implement traditional P2PKH send and receive functionality
    - [x] Tested in _regtest_ **and** _mainnet_ environments
- [x] Implement P2WPKH (Segwit) send and receive functionality
    - [x] Tested in _regtest_ **and** _mainnet_ environment
- [x] Implement P2TR (key-path) send and receive functionality
    - [x] Tested in _regtest_ **and** _mainnet_ environment
- [x] _(not directly relevant for this PR)_ Implement the Ordinals protocol and BRC20 scripts
    - Partially tested in _mainnet_ environment, but it's not functioning as expected. I'm currently debugging this. I initially wanted to test this in a testnet environment, but https://testnet.ordinals.com appears to be offline, and I couldn't obtain a clear answer from anyone on their Discord
- [x] Develop FFI bindings over relevant transaction Input/Output builders and transaction signing
    - [ ] Tested via FFI
    - [x] Wrap C++ structures over the FFI calls into Rust

The existing wallet-core library offers a lot of Bitcoin-related functionality such as fee estimation and message signing. We're retaining all of these for existing features in the iOS and Android apps (including traditional P2PKH Bitcoin transactions). Therefore, we're creating a separate C++ structure explicitly for Segwit/Taproot with similar APIs.

As for the code itself, several `TODO`'s and unhandled `.unwrap()`'s remain in the non-FFI exposed code (i.e., in the Ordinals/BRC20 sections). Rust API itself is likely to change. Btw, based on what I saw in the iOS codebase, it seems we only support P2PKH and (sending to) P2WPKH transactions, while others like P2PK and P2SH are not supported. Since we have Taproot now I think there's no longer a need to consider those, anyway.

### CODE REVIEW

File: rust/tw_bitcoin/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+}
+
+impl InputContext {
+    pub fn new(utxo: TxOut, point: OutPoint) -> Self {
+        InputContext {
+            previous_output: point,
+            value: Some(utxo.value),
+            script_pubkey: utxo.script_pubkey,
+            // Default value of `0xFFFFFFFF = 4294967295`.
+            sequence: Sequence::default(),
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/lib.rs:397]: Sequence can be different from 0 in some cases.
I guess we need to allow to set the sequence number for an input

File: rust/tw_bitcoin/src/ordinals.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+impl OrdinalsInscription {
+    /// Creates a new Ordinals Inscription ("commit stage").
+    pub fn new(
+        mime: &[u8],
+        data: &[u8],
+        recipient: Recipient<PublicKey>,
+    ) -> Result<OrdinalsInscription> {
+        let envelope = create_envelope(mime, data, recipient.public_key())?;
+        // TODO: In which cases is this `false`?
+        let merkle_root = envelope.spend_info.merkle_root().unwrap();
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ordinals.rs:41]: Just not to forget when this PR is r2r

File: rust/tw_bitcoin/src/ordinals.rs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+/// Do note that the `internal_key` can be different for each stage, but it
+/// could also be the same entity. Stage one, the `internal_key` is the
+/// recipient. Stage two, the `internal_key` is the claimer of the transaction
+/// (where the Inscription script is available in the Witness).
+fn create_envelope(mime: &[u8], data: &[u8], internal_key: PublicKey) -> Result<TaprootProgram> {
+    use bitcoin::opcodes::all::*;
+    use bitcoin::opcodes::*;
+
+    // TODO: Check overflow
+    let (op_push, size_buf) = get_op_push(mime.len() as u32)?;
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ordinals.rs:73]: Just not to forget when this PR is r2r

File: rust/tw_bitcoin/src/ordinals.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    // condition.
+    let spend_info = TaprootBuilder::new()
+        .add_leaf(0, script.clone())
+        .map_err(|_| Error::Todo)?
+        .finalize(
+            &secp256k1::Secp256k1::new(),
+            XOnlyPublicKey::from(internal_key.inner),
+        )
+        //.map_err(|_| Error::Todo)?;
+        .unwrap();
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ordinals.rs:122]: Maybe return an error?

File: rust/tw_bitcoin/src/ordinals.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+        .push_opcode(OP_PUSHBYTES_3)
+        .push_slice(b"ord")
+        .push_opcode(OP_PUSHNUM_1)
+        .push_slice(mime_buf)
+        .push_opcode(OP_PUSHBYTES_0)
+        .push_slice(data_buf)
+        .push_opcode(OP_ENDIF)
+        .into_script();
+
+    dbg!(&script);
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/ordinals.rs:108]: We should avoid printing anything to std.
So just not to forget when this PR is r2r:
https://doc.rust-lang.org/std/macro.dbg.html
> Note that the macro is intended as a debugging tool and therefore you should avoid having uses of it in version control for long periods (other than in tests and similar). Debug output from production code is better done with other facilities such as the [debug!](https://docs.rs/log/*/log/macro.debug.html) macro from the [log](https://crates.io/crates/log) crate.

File: rust/tw_bitcoin/src/recipient_impl.rs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+impl From<&KeyPair> for Recipient<PublicKey> {
+    fn from(keypair: &KeyPair) -> Self {
+        Recipient {
+            t: PublicKey::new(keypair.public_key()),
+        }
+    }
+}
+
+impl From<PublicKey> for Recipient<TweakedPublicKey> {
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/recipient_impl.rs:63]: Do we use `Recipient< TweakedPublicKey>`, `Recipient< WPubkeyHash>`, `Recipient< Taproot...>` somewhere?
If not, probably we could remove them?

File: rust/tw_bitcoin/src/tests/p2tr_script_path.rs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        .unwrap();
+
+    // Prepare BRC20 Deploy output for Alice.
+    let ticker = "qqqq".to_string();
+    let max = 1000;
+    let limit = None;
+    let decimals = None;
+
+    let brc20_deploy =
+        BRC20DeployInscription::new((&alice).into(), ticker.clone(), max, limit, decimals).unwrap();
```
Comment (satoshiotomakan) [inline rust/tw_bitcoin/src/tests/p2tr_script_path.rs:44]: Do we need to support BRC20 Deploy and Mint functionalities in Wallet Core?
cc @hewigovens

File: rust/tw_bitcoin/src/input/mod.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+pub use p2tr_script_path::*;
+pub use p2wpkh::*;
+
+#[derive(Debug, Clone)]
+pub enum TxInput {
+    P2PKH(TxInputP2PKH),
+    P2WPKH(TxInputP2WPKH),
+    P2TRKeyPath(TxInputP2TRKeyPath),
+    P2TRScriptPath(TxInputP2TRScriptPath),
+    NonStandard { ctx: InputContext },
```
[the remark this answers is not in the document]
Comment (lamafab) [reply rust/tw_bitcoin/src/input/mod.rs:20]: `P2TRScriptPath` is required for Ordinals/BRC20. `NonStandard` we can remove, I guess (for now)

[the remark this answers is not in the document]
Comment (lamafab) [reply rust/tw_bitcoin/src/input/mod.rs:20]: Regarding P2PKH, it is kind of required when you want to send BTC to Segwit/Taproot, but you have to claim a P2PKH output for spending.

File: rust/tw_bitcoin/src/ordinals.rs (hunk 5 above)
Comment (lamafab) [reply rust/tw_bitcoin/src/ordinals.rs:108]: Yep, forgot to remove this :)

File: rust/tw_bitcoin/src/tests/p2tr_script_path.rs (hunk 7 above)
Comment (lamafab) [reply rust/tw_bitcoin/src/tests/p2tr_script_path.rs:44]: I guess we do not, but once the underlying structure works those methods do not cost us anything.

File: rust/tw_bitcoin/src/recipient_impl.rs (hunk 6 above)
Comment (lamafab) [reply rust/tw_bitcoin/src/recipient_impl.rs:63]: It's kind of useful given that you can easily convert between `Keypair > PublicKey > PubkeyHash/WPubkeyHash`, given that each output type requires different type of data. I can probably simplify this, though.

Comment (lamafab) [reply rust/tw_bitcoin/src/recipient_impl.rs:63]: It's kind of useful given that you can easily convert between `Keypair > PublicKey > PubkeyHash/WPubkeyHash`, given that each output type requires different type of data. I can probably simplify this, though.

**Comment by lamafab** [comment]
Closing in favor of https://github.com/trustwallet/wallet-core/pull/3233

**Comment by lamafab** [comment]
@RomanDariicuk Can you please stop it with those cryptic replies? Clarify what you want or are trying to say.

**Pull request closed without merging**